app/src/main/java/io/appium/uiautomator2/handler/TouchEvent.java

```diff
@@ -0,0 +1,73 @@
+package io.appium.uiautomator2.handler;
+
+import android.graphics.Rect;
+import android.support.test.uiautomator.UiObjectNotFoundException;
+
+import com.jayway.jsonpath.JsonPath;
+
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import io.appium.uiautomator2.handler.request.SafeRequestHandler;
+import io.appium.uiautomator2.http.AppiumResponse;
+import io.appium.uiautomator2.http.IHttpRequest;
+import io.appium.uiautomator2.model.AndroidElement;
+import io.appium.uiautomator2.model.KnownElements;
+import io.appium.uiautomator2.server.WDStatus;
+import io.appium.uiautomator2.utils.Logger;
+
+public abstract class TouchEvent extends SafeRequestHandler {
+    protected int clickX, clickY;
+    private AndroidElement element;
+    private boolean isElement;
+
+    public TouchEvent(String mappedUri) {
+        super(mappedUri);
+    }
+
+    @Override
+    public AppiumResponse safeHandle(IHttpRequest request) {
+        String path = "$.params.";
+        try {
+            JSONObject json = getPayload(request);
+            if (json.has("id")) {
+                String id = json.getString("id");
+                element = KnownElements.getElementFromCache(id);
+
+                if (element != null)
+                    isElement = true;
+                if (isElement) {
```

Comment:
Yeah..Here  I have added the exception for the missed case. If we have element we will get the bounds else we can get the x and y  coordinates from the json.
```
          if (json.has("id")) {
                String id = json.getString("id");
                element = KnownElements.getElementFromCache(id);
                if (element != null) { 
                       final Rect bounds = element.getBounds();
                       clickX = bounds.centerX();
                       clickY = bounds.centerY();
                 } else {
                    throw new UiObjectNotFoundException("Touch Event Element doesn't exist");
                    }
                }               
            } else { // no element so extract x and y from params
                clickX = JsonPath.compile(path + "x").read(json.toString());
                clickY = JsonPath.compile(path + "y").read(json.toString());
            }
```